### PR TITLE
merge conflict bug fixes

### PR DESCRIPTION
**Pull request by ravishetye**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java [hunk 1]
```diff
@@ -135,6 +135,7 @@ private static AutogenVersionEnvironmentDiff getVersionDiff(
       AutogenVersionEnvironmentDiff a,
       AutogenVersionEnvironmentDiff b,
       AutogenVersionEnvironmentBlob c) {
+    if ((a == null && b == null) || (a != null && a.equals(b))) return null;
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:138]: addition like these  are resolving the actual issue reported in VR-4156

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java [hunk 2]
```diff
@@ -397,23 +400,28 @@ private static AutogenCodeDiff getCodeConflictBlob(
...
-            .setPath(
-                new AutogenPathDatasetComponentDiff()
-                    .setA(a == null || a.getPath() == null ? null : a.getPath().getB())
-                    .setB(b == null || b.getPath() == null ? null : b.getPath().getB())
-                    .setC(c == null ? null : c.getPath())
-                    .setStatus(AutogenDiffStatusEnumDiffStatus.fromProto(DiffStatus.CONFLICTED))));
+                    c == null ? null : c.getGitRepo()));
+    if (((a == null || a.getPath() == null) && (b == null || b.getPath() == null))
+        || (a != null && a.equals(b))) return Utils.removeEmpty(diff);
+    diff.setPath(
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:413]: there was a bug where PathDataset would appear as a conflict when not expected

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/DiffMerger.java [hunk 3]
```diff
@@ -70,7 +70,10 @@ else if (!elA.contains(diffB)) {
                         // TODO: error otherwise
                         if (diffA != null && diffB != null) {
                           if (!elA.contains(diffA)) { // diff is applied on a state different from A
-                            conflictKeys.add(key);
+                            if (!elA.contains(
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/DiffMerger.java:73]: when the end modification was exactly the same we would still call a conflict , this fixes that case

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 2 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:413]: I'm not sure I follow what this does and why it's an issue only for notebook paths.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java [hunk 4]
```diff
@@ -397,23 +400,28 @@ private static AutogenCodeDiff getCodeConflictBlob(
...
+        new AutogenPathDatasetComponentDiff()
+            .setA(a == null || a.getPath() == null ? null : a.getPath().getB())
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:415]:
```suggestion
            .setA(a == null || a.getPath() == null ? null : a.getPath().getA())
```
Maybe?

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 1 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:138]: Could you expand on why they resolve?

**Review by conradoverta — COMMENTED** [review]
I am concerned about fixing just for the cases that were reported and leaving similar latent issues behind. How could we verify that is not the case? Is there a way for us to build a property test to verify the new properties we're creating?

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 1 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:138]: during the merge attempt, we collect the locations of the conflicts. A location could have a python env blob , which has  the same python version but a conflict in requirements . this code would remove the BLOBs which are not in conflict mainly by `a.equals(b)`.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 2 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:413]: bug was the code setting a AutogenPathDatasetComponentDiff without validation. so even if a,b,c are null a stray Status.conflict was present in the conflicts. this is not related directly to the bug you reported , but was identified during the testing

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 4 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:415]: no these are the diffs of the commits being merged so get B is correct , getA would be the common parent

**Comment by ravishetye** [comment]
> I am concerned about fixing just for the cases that were reported and leaving similar latent issues behind. How could we verify that is not the case? Is there a way for us to build a property test to verify the new properties we're creating?

actually this is not fixing just the reported cases, there are multiple issues which were identified and fixed.

That aside , I agree the merge tests can be made more targetted as well. Will address the issue in another PR and track it in jira

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 4 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:415]: Sg

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java (hunk 2 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:413]: Wouldn't this also be an issue for the Git part of the blob?

Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/diff/ConflictGenerator.java:413]: the git part is handled in getGitDiff

**Review by conradoverta — APPROVED** [review]
LGTM

**Pull request merged**